File: content/develop/use-cases/time-series-dashboard/dotnet/RedisTimeSeriesStore.cs
using System.Text.Json.Serialization;
using NRedisStack;
using NRedisStack.DataTypes;
using NRedisStack.Literals.Enums;
using NRedisStack.RedisStackCommands;
using StackExchange.Redis;

namespace TimeSeriesDashboardDemo;

public sealed class RedisTimeSeriesStore
{
    public const int SampleIntervalMs = 500;
    public const int WindowMs = 12_000;
    public const int BucketMs = 3_000;
    public const int RetentionMs = 12_000;

    private readonly ITimeSeriesCommandsAsync _ts;
    private readonly IReadOnlyList<SensorDefinition> _sensors;

    public RedisTimeSeriesStore(IDatabase db, IReadOnlyList<SensorDefinition> sensors)
    {
        _ts = db.TS();
        _sensors = sensors;
    }

    public async Task EnsureSchemaAsync()
    {
        foreach (var sensor in _sensors)
        {
            var labels = sensor.Labels
                .Select(label => new TimeSeriesLabel(label.Key, label.Value))
                .ToList();

            try
            {
                await _ts.CreateAsync(sensor.Key, retentionTime: RetentionMs, labels: labels);
            }
            catch (RedisServerException ex) when (ex.Message.Contains("key already exists", StringComparison.OrdinalIgnoreCase))
            {
            }
        }
    }

    public Task AddSamplesAsync(IReadOnlyList<SensorSample> samples)
    {
        if (samples.Count == 0)
        {
            return Task.CompletedTask;
        }

        var timestampMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var sequence = samples
            .Select(sample => (sample.Sensor.Key, (TimeStamp)timestampMs, sample.Value))
            .ToList();

        return _ts.MAddAsync(sequence);
    }

    public async Task<DashboardSnapshot> DashboardSnapshotAsync()
    {
        var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var startMs = nowMs - WindowMs;
        var aggregateStartMs = startMs - BucketMs;
        var sensorSnapshots = new List<SensorSnapshot>(_sensors.Count);

        foreach (var sensor in _sensors)
        {
            var rawPoints = await RangeQueryAsync(sensor, startMs, nowMs);
            var minPoints = await AggregateQueryAsync(sensor, aggregateStartMs, nowMs, TsAggregation.Min);
            var maxPoints = await AggregateQueryAsync(sensor, aggregateStartMs, nowMs, TsAggregation.Max);
            var avgPoints = await AggregateQueryAsync(sensor, aggregateStartMs, nowMs, TsAggregation.Avg);
            var latest = await LatestQueryAsync(sensor);

            var minByBucket = IndexPoints(minPoints);
            var maxByBucket = IndexPoints(maxPoints);
            var avgByBucket = IndexPoints(avgPoints);

            var firstBucketStart = (startMs / BucketMs) * BucketMs;
            if (firstBucketStart > startMs)
            {
                firstBucketStart -= BucketMs;
            }

            var lastBucketStart = (nowMs / BucketMs) * BucketMs;
            var buckets = new List<BucketSnapshot>();

            for (var bucketStart = firstBucketStart; bucketStart <= lastBucketStart; bucketStart += BucketMs)
            {
                buckets.Add(new BucketSnapshot
                {
                    Start = bucketStart,
                    End = bucketStart + BucketMs,
                    Avg = avgByBucket.GetValueOrDefault(bucketStart),
                    Min = minByBucket.GetValueOrDefault(bucketStart),
                    Max = maxByBucket.GetValueOrDefault(bucketStart)
                });
            }

            sensorSnapshots.Add(new SensorSnapshot
            {
                SensorId = sensor.SensorId,
                Zone = sensor.Zone,
                Unit = sensor.Unit,
                Latest = latest,
                RawPoints = rawPoints,
                Buckets = buckets
            });
        }

        return new DashboardSnapshot
        {
            Now = nowMs,
            WindowMs = WindowMs,
            BucketMs = BucketMs,
            SampleIntervalMs = SampleIntervalMs,
            RetentionMs = RetentionMs,
            Sensors = sensorSnapshots
        };
    }

    private async Task<List<PointSnapshot>> RangeQueryAsync(SensorDefinition sensor, long startMs, long endMs)
    {
        var tuples = await _ts.RangeAsync(sensor.Key, startMs, endMs);
        return ToPointSnapshots(tuples);
    }

    private async Task<List<PointSnapshot>> AggregateQueryAsync(
        SensorDefinition sensor,
        long startMs,
        long endMs,
        TsAggregation aggregation)
    {
        var tuples = await _ts.RangeAsync(
            sensor.Key,
            startMs,
            endMs,
            align: 0L,
            aggregation: aggregation,
            timeBucket: BucketMs);

        return ToPointSnapshots(tuples);
    }

    private async Task<LatestValueSnapshot?> LatestQueryAsync(SensorDefinition sensor)
    {
        try
        {
            var tuple = await _ts.GetAsync(sensor.Key);
            return tuple is null
                ? null
                : new LatestValueSnapshot { Timestamp = tuple.Time, Value = tuple.Val };
        }
        catch (RedisServerException ex) when (ex.Message.Contains("does not exist", StringComparison.OrdinalIgnoreCase))
        {
            return null;
        }
    }

    private static List<PointSnapshot> ToPointSnapshots(IEnumerable<TimeSeriesTuple> tuples) =>
        tuples
            .Select(tuple => new PointSnapshot { Timestamp = tuple.Time, Value = tuple.Val })
            .ToList();

    private static Dictionary<long, double?> IndexPoints(IEnumerable<PointSnapshot> points)
    {
        return points.ToDictionary(point => point.Timestamp, point => (double?)point.Value);
    }
}

public sealed class PointSnapshot
{
    [JsonPropertyName("timestamp")]
    public long Timestamp { get; init; }

    [JsonPropertyName("value")]
    public double Value { get; init; }
}

public sealed class BucketSnapshot
{
    [JsonPropertyName("start")]
    public long Start { get; init; }

    [JsonPropertyName("end")]
    public long End { get; init; }

    [JsonPropertyName("avg")]
    public double? Avg { get; init; }

    [JsonPropertyName("min")]
    public double? Min { get; init; }

    [JsonPropertyName("max")]
    public double? Max { get; init; }
}

public sealed class LatestValueSnapshot
{
    [JsonPropertyName("timestamp")]
    public long Timestamp { get; init; }

    [JsonPropertyName("value")]
    public double Value { get; init; }
}

public sealed class SensorSnapshot
{
    [JsonPropertyName("sensor_id")]
    public required string SensorId { get; init; }

    [JsonPropertyName("zone")]
    public required string Zone { get; init; }

    [JsonPropertyName("unit")]
    public required string Unit { get; init; }

    [JsonPropertyName("latest")]
    public LatestValueSnapshot? Latest { get; init; }

    [JsonPropertyName("raw_points")]
    public required List<PointSnapshot> RawPoints { get; init; }

    [JsonPropertyName("buckets")]
    public required List<BucketSnapshot> Buckets { get; init; }
}

public sealed class DashboardSnapshot
{
    [JsonPropertyName("now")]
    public long Now { get; init; }

    [JsonPropertyName("window_ms")]
    public long WindowMs { get; init; }

    [JsonPropertyName("bucket_ms")]
    public long BucketMs { get; init; }

    [JsonPropertyName("sample_interval_ms")]
    public long SampleIntervalMs { get; init; }

    [JsonPropertyName("retention_ms")]
    public long RetentionMs { get; init; }

    [JsonPropertyName("sensors")]
    public required List<SensorSnapshot> Sensors { get; init; }
}

File: content/develop/use-cases/time-series-dashboard/php/RedisTimeSeriesStore.php
<?php

use Predis\Client as PredisClient;

final class RedisTimeSeriesStore
{
    public const SAMPLE_INTERVAL_MS = 500;
    public const WINDOW_MS = 12000;
    public const BUCKET_MS = 3000;
    public const RETENTION_MS = 12000;

    /** @var list<SensorDefinition> */
    private array $sensors;

    /**
     * @param list<SensorDefinition> $sensors
     */
    public function __construct(
        private readonly PredisClient $redis,
        array $sensors
    ) {
        $this->sensors = $sensors;
    }

    public function ensureSchema(): void
    {
        foreach ($this->sensors as $sensor) {
            $args = ['TS.CREATE', $sensor->key(), 'RETENTION', (string) self::RETENTION_MS, 'LABELS'];
            foreach ($sensor->labels() as $label => $value) {
                $args[] = $label;
                $args[] = $value;
            }

            try {
                $this->redis->executeRaw($args);
            } catch (\Throwable $exception) {
                if (stripos($exception->getMessage(), 'key already exists') === false) {
                    throw $exception;
                }
            }
        }
    }

    /**
     * @param list<array{sensor: SensorDefinition, value: float}> $samples
     */
    public function addSamples(array $samples, ?int $timestampMs = null): void
    {
        if ($samples === []) {
            return;
        }

        $timestampMs ??= (int) floor(microtime(true) * 1000);
        $args = ['TS.MADD'];
        foreach ($samples as $sample) {
            $args[] = $sample['sensor']->key();
            $args[] = (string) $timestampMs;
            $args[] = $this->formatFloat($sample['value']);
        }

        $this->redis->executeRaw($args);
    }

    /**
     * @return array<string, mixed>
     */
    public function dashboardSnapshot(): array
    {
        $nowMs = (int) floor(microtime(true) * 1000);
        $startMs = $nowMs - self::WINDOW_MS;
        $aggregateStartMs = $startMs - self::BUCKET_MS;
        $sensorPayload = [];

        foreach ($this->sensors as $sensor) {
            $rawPoints = $this->range($sensor, $startMs, $nowMs);
            $minPoints = $this->aggregate($sensor, $aggregateStartMs, $nowMs, 'min');
            $maxPoints = $this->aggregate($sensor, $aggregateStartMs, $nowMs, 'max');
            $avgPoints = $this->aggregate($sensor, $aggregateStartMs, $nowMs, 'avg');
            $latest = $this->latest($sensor);

            $minByBucket = $this->indexByTimestamp($minPoints);
            $maxByBucket = $this->indexByTimestamp($maxPoints);
            $avgByBucket = $this->indexByTimestamp($avgPoints);

            $firstBucketStart = intdiv($startMs, self::BUCKET_MS) * self::BUCKET_MS;
            if ($firstBucketStart > $startMs) {
                $firstBucketStart -= self::BUCKET_MS;
            }
            $lastBucketStart = intdiv($nowMs, self::BUCKET_MS) * self::BUCKET_MS;

            $buckets = [];
            for ($bucketStart = $firstBucketStart; $bucketStart <= $lastBucketStart; $bucketStart += self::BUCKET_MS) {
                $buckets[] = [
                    'start' => $bucketStart,
                    'end' => $bucketStart + self::BUCKET_MS,
                    'avg' => $avgByBucket[$bucketStart] ?? null,
                    'min' => $minByBucket[$bucketStart] ?? null,
                    'max' => $maxByBucket[$bucketStart] ?? null,
                ];
            }

            $sensorPayload[] = [
                'sensor_id' => $sensor->sensorId,
                'zone' => $sensor->zone,
                'unit' => $sensor->unit,
                'latest' => $latest,
                'raw_points' => $rawPoints,
                'buckets' => $buckets,
            ];
        }

        return [
            'now' => $nowMs,
            'window_ms' => self::WINDOW_MS,
            'bucket_ms' => self::BUCKET_MS,
            'sample_interval_ms' => self::SAMPLE_INTERVAL_MS,
            'retention_ms' => self::RETENTION_MS,
            'sensors' => $sensorPayload,
        ];
    }

    /**
     * @return list<array{timestamp: int, value: float}>
     */
    private function range(SensorDefinition $sensor, int $startMs, int $endMs): array
    {
        $response = $this->redis->executeRaw([
            'TS.RANGE',
            $sensor->key(),
            (string) $startMs,
            (string) $endMs,
        ]);

        return $this->parsePoints($response);
    }

    /**
     * @return list<array{timestamp: int, value: float}>
     */
    private function aggregate(SensorDefinition $sensor, int $startMs, int $endMs, string $aggregation): array
    {
        $response = $this->redis->executeRaw([
            'TS.RANGE',
            $sensor->key(),
            (string) $startMs,
            (string) $endMs,
            'ALIGN',
            '0',
            'AGGREGATION',
            $aggregation,
            (string) self::BUCKET_MS,
        ]);

        return $this->parsePoints($response);
    }

    /**
     * @return array{timestamp: int, value: float}|null
     */
    private function latest(SensorDefinition $sensor): ?array
    {
        try {
            $response = $this->redis->executeRaw(['TS.GET', $sensor->key()]);
        } catch (\Throwable $exception) {
            if (stripos($exception->getMessage(), 'does not exist') !== false) {
                return null;
            }

            throw $exception;
        }

        if (!is_array($response) || count($response) < 2) {
            return null;
        }

        return [
            'timestamp' => (int) $response[0],
            'value' => (float) $response[1],
        ];
    }

    /**
     * @param mixed $response
     * @return list<array{timestamp: int, value: float}>
     */
    private function parsePoints(mixed $response): array
    {
        if (!is_array($response)) {
            return [];
        }

        $points = [];
        foreach ($response as $row) {
            if (!is_array($row) || count($row) < 2) {
                continue;
            }

            $points[] = [
                'timestamp' => (int) $row[0],
                'value' => (float) $row[1],
            ];
        }

        return $points;
    }

    /**
     * @param list<array{timestamp: int, value: float}> $points
     * @return array<int, float>
     */
    private function indexByTimestamp(array $points): array
    {
        $indexed = [];
        foreach ($points as $point) {
            $indexed[$point['timestamp']] = $point['value'];
        }

        return $indexed;
    }

    private function formatFloat(float $value): string
    {
        return rtrim(rtrim(number_format($value, 6, '.', ''), '0'), '.');
    }
}
